src/SFML/Graphics/Texture.cpp

```diff
@@ -854,6 +861,10 @@ void Texture::bind(const Texture* texture, CoordinateType coordinateType)
 
     if (texture && texture->m_texture)
     {
+        // When debugging, ensure that the texture name is valid
+        assert((glIsTexture(texture->m_texture) == GL_TRUE) &&
+               "Texture to be bound is invalid, check if the texture is still being used after it has been destroyed");
+
```

Comment:
I've thought on this some more. I want to make sure I have a correct understanding of what's happening.

_In the case the texture is still alive:_
No UB is invoked. `texture->m_texture` returns a well defined value and we can be confident `glIsTexture(texture->m_texture)` returns `GL_TRUE`. This assertion does not fire. All is well. For users who write correct programs the addition of this assertion will not cause them any problems.

_In the case the texture is not alive:_
UB is invoked. `texture->m_texture` returns an undefined value. Perhaps it returns the value assigned to it in the destructor. I'm inclined to believe this is what will typically happen but we can't be sure of that. If the expected value is returned by the expression `texture->m_texture` then the assertion correctly fails. This is a good outcome. However, what is a different value is returned? There seems to be a theoretical possibility that `texture->m_texture` returns a value that causes the assertion to _not_ fail. In this case it's as if the assertion was never added which I suppose isn't necessarily a problem. Statistically this seems unlikely so I'm not especially worried about this happening. If this does happen then, well, the users code was wrong to begin with so the false negative means they're no worse off than before.

My scarier thought is what else is possible when invoking UB in this case? The compiler is going to assume the texture is alive and will make transformations based on that assumption. What if the compiler decides to delete some code under the assumption that the texture is alive? What code will get deleted? How will that change the semantics of the users's program? Can we be sure that won't happen? It's hard to put bounds on what may happen when UB is invoked. We can typically assume some things will or won't happen but those assumptions are not guarantees and are prone to change as the language and compilers evolve. Reasoning about UB isn't a totally useless exercise but it's damn hard.